### PR TITLE
Dev/robin/9780 veracity update-replica command

### PR DESCRIPTION
**Pull request by robinbryce**
The update-replica command supports efficient independent 3rd party verification of datatrails logs

When combined with the output of the watch command, this allows anyone to safely maintain a replica, or partial replica, of a datatrails transparency log.

The remote log is first verified in its own terms, then any additional content in the remote log is verified to be consistent with the most recent localy replicated state for the same tenant.

### CODE REVIEW

File: verifyincluded.go [hunk 1]
```diff
@@ -29,14 +29,18 @@ func proofPath(proof [][]byte) string {
...
 
-func verifyEvent(event *logverification.VerifiableEvent, massifHeight uint8, massifReader MassifReader, forTenant string) ([][]byte, error) {
+func verifyEvent(
+	event *logverification.VerifiableEvent, massifHeight uint8, massifReader MassifReader,
+	forTenant string,
+	tenantLogPath string,
+) ([][]byte, error) {
 
 	// Get the mmrIndex from the request and then compute the massif
 	// it implies based on the massifHeight command line option.
```
Comment (robinbryce) [inline verifyincluded.go:39]: Uh, I think I need to rebase this

**Review by henry739 — COMMENTED** [review]
Some minor points - looking fantastic though

File: verifyconsistency.go [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+	ErrFailedToCreateReplicaDir       = errors.New("failed to create a directory needed for local replication")
+	ErrRequiredOption                 = errors.New("a required option was not provided")
+	ErrRemoteLogTruncated             = errors.New("the local replica indicates the remote log has been truncated")
+	ErrRemoteLogInconsistentRootState = errors.New("the local replica root state disagrees with the remote")
+)
+
+// NewVerifyConsistencyCmd updates a local replica of a remote log, verifying the mutual consistency of the two before making any changes.
+//
+//nolint:gocognit
+func NewVerifyConsistencyCmd() *cli.Command {
```
[the remark this answers is not in the document]
Comment (robinbryce) [reply verifyconsistency.go:31]: yeah, I agree. I wanted to call it something like `veracity verify-replica` or something that made 'update and verify' the primary sense. maybe `veracity update-trusted`. The original ticket said 'verify consistency` so I left it as that. But I think we will _also_ want a verify-consistency that just checks a single remote (or local) massif and its seal against a provided consistency proof. So yes, I'm in favour of a better name and its about 30 minutes to change pre merge

File: tests/verifyconsistency/verifyconsistency_azurite_test.go [hunk 3]
```diff
@@ -0,0 +1,530 @@
...
+	// check the 0'th massifs and seals was replicated (by the first run of veractity)
+	expectMassifFile := filepath.Join(replicaDir, massifs.ReplicaRelativeMassifPath(tenantId0, 0))
+	s.FileExistsf(expectMassifFile, "the replicated massif should NOT exist")
+	expectSealFile := filepath.Join(replicaDir, massifs.ReplicaRelativeSealPath(tenantId0, 0))
+	s.FileExistsf(expectSealFile, "the replicated seal should NOT exist")
+
+	// check the massifs from the second veracity run were NOT replicated
+	for i := uint32(1); i < massifCount; i++ {
+		expectMassifFile := filepath.Join(replicaDir, massifs.ReplicaRelativeMassifPath(tenantId0, i))
+		s.NoFileExistsf(expectMassifFile, "the replicated massif should exist")
```
[the remark this answers is not in the document]
Comment (robinbryce) [reply tests/verifyconsistency/verifyconsistency_azurite_test.go:406]: I found a few that were the wrong way round.

FileExists(filename, ...) fails if the file does not exist, so the message should indicate the expectation that it does exist. NoFileExists is the inversion of that. A bunch of lazy copy paste made some of those inconsistent. should be good now

File: verifyincluded.go [hunk 4]
```diff
@@ -112,9 +119,12 @@ Note: for publicly attested events, or shared protected events, you must use --t
 				cmd.log.Infof(m, args...)
 			}
 
-			log("verifying events dir: %s", cCtx.String("logdir"))
-
-			verifiableEvents, err := stdinToVerifiableEvents()
+			var verifiableEvents []logverification.VerifiableEvent
+			if cCtx.Args().Len() > 0 {
```
[the remark this answers is not in the document]
Comment (robinbryce) [reply verifyincluded.go:123]: I think it is helpful ux to be permissive of additional arguments. but sure, that could be argued either way. I've added a comment to convey the intention

File: verifyconsistency.go [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+		// state we do not attempt to automaticaly back fill. This is so it is
+		// possible to ensure a bounded number of requests (and copies) in
+		// situations where verifiers have been off line for a long time.
+		//
+		// Explicit back filling is always possible by running with --ancestor=0
+		// or by running the command multiple times with increasing --massif
+
+		// With that accounted for, we then want to avoid re-verifying the state
+		// we have already verified locally, hence the use of `max`.
+		// We make no attempt to deal with "sparse" collections of massifs, the
```
[the remark this answers is not in the document]
Comment (robinbryce) [reply verifyconsistency.go:317]: Correct. And the way to force fill in the gaps is to set `--ancestors=0` which means "make sure all are present", and this is also the default behaviour if  `--ancestors` is not set. THe help text for --ancestors ends with  "If set to 0, a full replica is requested."

This is something I intend to make pretty clear in the user docs

File: verifyconsistency.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (robinbryce) [reply verifyconsistency.go:31]: I've raised this as a question on the feature channel. I'll apply the result of that conversation to this PR

**Pull request merged**